tools/pathwaymatcher/pathwaymatcher.xml

```diff
@@ -0,0 +1,332 @@
+
+            #if $genetic_variants.input_vcf:
+                -t vcf -i "${genetic_variants.input_vcf}"
+            #end if
+
+
+            ## GENES
+
+            #if $genes.input_gene:
+                -t gene -i "${genes.input_gene}"
+            #end if
+
+            ## PEPTIDES
+
+##            #if $peptides.input_peptide:
+##                -t peptide -i "${peptides.input_peptide}"
+##            #end if
+
+##            #if $peptides.input_modifiedpeptide:
+##                -t modifiedpeptide -i "${peptides.input_modifiedpeptide}"
+##            #end if
+
+##            #if $peptides.ptm_range:
+##                -r "${peptides.ptm_range}"
+##            #end if
+
+
+            ## PROTEINS
+
+            #if $proteins.input_uniprot:
+                -t uniprot -i "${proteins.input_uniprot}"
+            #end if
+
+            #if $proteins.input_ensembl:
+                -t ensembl -i "${proteins.input_ensembl}"
+            #end if
+
+
+            ## PROTEOFORMS
+
+            #if $proteoforms.input_proteoforms:
+                #if $proteoforms.proteoform_match_criteria:
+                  -t proteoform -m "${proteoforms.proteoform_match_criteria}" -i "${proteoforms.input_proteoforms}"
+                #else:
+                  -t proteoform -i "${proteoforms.input_proteoforms}"
+                #end if
+            #end if
+
+
+            ## OUTPUT OPTIONS
+
+            #if $output_options.proteoform_match_criteria:
+                -tlp
+            #end if
+
+            #set $output_graphs_list = str($output_options.output_graphs).split(',')
+
+            #if 'gg' in $output_graphs_list:
+                -gg
+            #end if
+
+            #if 'gu' in $output_graphs_list:
+                -gu
+            #end if
+
+            #if 'gp' in $output_graphs_list:
+                -gp
+            #end if
+
+        2>> $temp_stderr);
+
+        ## We create a zip file to contain graphs files.
+        #if $output_options.output_graphs:
+            touch "emptyfile.txt";
+            zip "graphs.zip" "emptyfile.txt";
+            zip -d "graphs.zip" "emptyfile.txt";
+        #end if
+
+        #if 'gg' in $output_graphs_list:
+            zip -u "graphs.zip" "geneExternalEdges.tsv" "geneInternalEdges.tsv" "geneVertices.tsv";
+        #end if
+
+        #if 'gu' in $output_graphs_list:
+            zip -u "graphs.zip" "proteinExternalEdges.tsv" "proteinInternalEdges.tsv" "proteinVertices.tsv";
+        #end if
+
+        #if 'gp' in $output_graphs_list:
+            zip -u "graphs.zip" "proteoformExternalEdges.tsv" "proteoformInternalEdges.tsv" "proteoformVertices.tsv";
+        #end if
+
+
+        exit_code_for_galaxy=\$?;
+        cat $temp_stderr 2>&1;
+        (exit \$exit_code_for_galaxy)
+]]>
+    </command>
+    <inputs>
+
+        <section name="genetic_variants" expanded="true" title="Genetic variants">
+
+          <param format="txt" name="input_rsid" type="data" label="SNP rsId list" optional="true"
```

Comment:
As far as I can see all inputs are optional? This mean a tool can be submitted without any input?

Reply:
Yes, in theory the tool can be submitted without any input, but of course it will give an error.
At least one input is necessary to get results, but I'm not sure about how to do that...

Reply:
what about a repeat?

Reply:
repeat? I don't understand sorry

Reply:
https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-inputs-repeat

You add a repeat and people select the type + file. And can add up to X repeat tuples.

Reply:
Thank you, I will check it.

Reply:
Um... I think I'm not understanding the repeat clause. Adding the repeat clause you can still press the execute button without selecting any file. Even after pressing over "insert input files" you can still press the Execute button without adding files at all. And pressing again "insert input files" duplicates the entire interface without necessity, since with the original structure you can add extra files directly without any extra step...

Reply:
What about something like this:

```
<repeat name="samples" title="Samples" min="1">
  <param argument="-i" type="data" format="data" label="UniProt Accession list"
       help="File with a one Uniprot Accession [4] in each line."/>
  <param argument="-t" type="select" label="type">
       <option value="rsid">rsid</option>
       .....
  </param>
</repeat>
```

Reply:
Yes, I did it. Although I specify min="1" you can still press on "execute" without adding any file... and the duplicated visualisation of input params when pressing on "insert input files" I think is not a good solution...

Reply:
You need to remove the optional=True from your data param.

Reply:
Um... it still doesn`t do what I need. If I remove "optional=True" of all parameters, what happens is all of them have one value chosen by default without the possibility of removing it.
What it should do is to have them all empty by default (as only one or some of them will be used) and not allow the user to execute the tool without choosing at least one value for one param...

Reply:
Carlos can you show me the code? I think I misunderstood what you want?

Reply:
Sure, I simplified it to this to check how "repeat" works:

'''
<inputs>

        <repeat name="input_files" title="Input files" min="1">

              <param format="txt" argument="-i" name="input_rsid" type="data" label="SNP rsId list"
                help="The file contains one rsid identifier as defined in dbSNP[1] on each row.
                The list must be ordered by chromosome and base pair (bp). The list must not have duplicates.
                All rsids must appear in the human assembly GRCh37.p13. "/>

              <param format="txt" argument="-t" name="input_gene" type="data" label="Genes"
                help="File with a one gene name in each line. Genes follow the HUGO gene nomenclature[3]."/>

        </repeat>

    </inputs>
'''

Reply:
So this is how I understand your use case you have X tuples of (-i and -t) parameters. You can add 1-X different tuples, right? But at least one tuple. So you start with one repeat. Yes you need to choose one file and one -i parameter (which I would do as select, why is this a file in your case?) The file is not allowed to be empty, because it is the first repeat. The -i is also not to be allowed to be empty because the tuple needs to be complete.
If you now choose to enter a second file you add a new repeat block and start all over again. Again if you have a second repeat blog, you need to enter -i & -t otherwise it is not complete, or remove the second repeat block. No?

Reply:
Um... I think I understood now... I have to think about it

Reply:
What do you think about this example?

'''
        <repeat name="input_types" title="Input Types" min="1">
            <conditional name="input_type">
                  <param name="input_type_selector" type="select" label="Input type"
                      help="">
                      <option value="rsid">Genetic variants - SNP rsId list</option>
                      <option value="chrbp">Genetic variants - Chromosomes and base pairs</option>
                      <option value="vcf">Genetic variants - Variant Call Format Specification</option>
                      <option value="peptide">Peptides - Simple list</option>
                     
                  </param>
                  <when value="rsid">
                      <param format="txt" name="input_rsid" type="data" label="SNP rsId list"
                        help="The file contains one rsid identifier as defined in dbSNP[1] on each row.
                        The list must be ordered by chromosome and base pair (bp). The list must not have duplicates.
                        All rsids must appear in the human assembly GRCh37.p13. "/>
                  </when>

                  <when value="chrbp">
                      <param format="txt" name="input_chrbp" type="data" label="Chromosomes and base pairs"
                        help="Genetic variants can also be represented using the chromosome and the base pair numbers.
                        The input should be sorted by chromosome number and then by base pair.  "/>
                  </when>

                  <when value="vcf">
                        <param format="txt" name="input_vcf" type="data" label="Variant Call Format Specification"
                          help="The input follows the Variant Call Format Specification[2] v4.3.
                          It also allows the possibility to specify only the first 4 columns in the data section of the file:
                          CHROM, POS, ID, REF.  "/>
                  </when>

                  <when value="peptide">
                        <param format="txt" name="input_peptide" type="data" label="Simple list"
                          help="File with a one peptide sequence in each line. You also need to choose a FASTA file to search them against it."/>

                        <param name="ptm_range" type="integer" value="0" label="PTM position range" optional="true"
                            help="Plus minus positions for the same PTM site."/>
                  </when>


            </conditional>
          
        </repeat>

'''

Main changes:
1) Select clause replaced by a conditional one. It allows to add necessary extra (and optional) options to some inputs (ie: peptides, for choosing PTM range)

2) It allows a more customised second field for the input files (with help and others)